src/mainwindow.cpp

```diff
@@ -1317,6 +1337,113 @@ void MainWindow::checkForUpdates (const bool clicked) {
     m_updater.checkForUpdates (false);
 }
 
+/**
+ * Called when the "Import Notes" menu button is clicked. this function will
+ * prompt the user to select a file, attempt to load the file, and update the DB
+ * if valid.
+ * The user is presented with a dialog box if the upload/import fails for any reason.
+ *
+ * @brief MainWindow::importNotesFile
+ * @param clicked
+ */
+void MainWindow::importNotesFile (const bool clicked) {
+    Q_UNUSED (clicked);
+    QString fileName = QFileDialog::getOpenFileName(this,
+            tr("Open Notes Backup File"), "",
+            tr("Notes Backup File (*.nbk)"));
+
+    if (fileName.isEmpty()) {
+        return;
+    } else {
+        QFile file(fileName);
+        if (!file.open(QIODevice::ReadOnly)) {
+            QMessageBox::information(this, tr("Unable to open file"), file.errorString());
+            return;
+        }
+        QList<NoteExport> noteList;
+        QDataStream in(&file);
+        in.setVersion(QDataStream::Qt_5_6);
+        try {
+            in >> noteList;
+        } catch (...) {
+            // Any exception deserializing will result in an empty note list and  the user will be notified
+        }
+
+        if (noteList.isEmpty()) {
+            QMessageBox::information(this, tr("Invalid file"), "Please select a valid notes export file");
+            return;
+        }
+
+        QProgressDialog* pd = new QProgressDialog("Importing Notes...", "", 0, 0, this);
+        pd->setCancelButton(0);
+        pd->setWindowFlags(Qt::Window | Qt::FramelessWindowHint);
+        pd->setMinimumDuration(0);
+        pd->show();
+
+        setButtonsAndFieldsEnabled(false);
+
+        QFutureWatcher<void>* watcher = new QFutureWatcher<void>(this);
+        connect(watcher, &QFutureWatcher<void>::finished, this, [&, pd](){
+            pd->deleteLater();
+
+            setButtonsAndFieldsEnabled(true);
+
+            m_noteModel->clearNotes();
+            loadNotes();
+            createNewNoteIfEmpty();
+            selectFirstNote();
+        });
+
+        QFuture<void> migration = QtConcurrent::run(this, &MainWindow::importNotes, noteList);
+        watcher->setFuture(migration);
+    }
+}
+
+void MainWindow::importNotes(QList<NoteExport> noteList) {
+
+    QList<QFuture<void>> futures;
+    for (int i = 0; i < noteList.size(); ++i) {
+        futures <<  QtConcurrent::run(this, &MainWindow::importNote, noteList[i]);
+    }
+    for (int i = 0; i < futures.size(); ++i) {
+        while (!futures[i].isFinished()) {
+            // block until all threads complete
+        }
+    }
+}
```

Comment:
you can use [void QtConcurrent::blockingMap(Sequence &sequence, MapFunction function)](http://doc.qt.io/qt-5/qtconcurrent.html#blockingMap) instead of the two loops